File: ios/Yappy/Features/Chat/MentionsScreen.swift
import SwiftUI

/// Mentions and platform notices share one chronological inbox.
/// The existing route name is kept so saved navigation remains valid.
struct MentionsScreen: View {
    @Environment(\.neu) private var colors
    @EnvironmentObject private var container: AppContainer

    let onBack: () -> Void
    /// Opens the room *at* the message, not merely at the bottom of it.
    let onOpenMessage: (String, Int64) -> Void
    let onOpenGroup: (String) -> Void
    let onOpenProfile: (String) -> Void

    @StateObject private var model = NotificationInboxModel()

    var body: some View {
        VStack(spacing: 0) {
            HStack(spacing: 12) {
                NeuIconButton(systemName: "chevron.left", label: "Back", action: onBack)
                Text("Notifications")
                    .font(YappyFont.headlineSmall)
                    .foregroundStyle(colors.textPrimary)
                Spacer(minLength: 0)
            }
            .padding(.horizontal, 16)
            .padding(.vertical, 12)

            List {
                if !model.loaded && model.loading {
                    ProgressView("Loading…")
                        .frame(maxWidth: .infinity)
                } else if model.loaded && model.entries.isEmpty && model.error == nil {
                    Text("Nothing yet. Mentions, verification updates, affiliations, and new roles land here.")
                        .foregroundStyle(colors.textTertiary)
                        .padding(.vertical, 24)
                }
                ForEach(model.entries) { entry in
                    Group {
                        switch entry {
                        case .mention(let mention): row(mention)
                        case .notice(let notice):
                            NotificationRow(entry: notice, onOpenGroup: onOpenGroup,
                                            onOpenProfile: onOpenProfile, onOpenMessage: onOpenMessage)
                        }
                    }
                    .listRowInsets(EdgeInsets(top: 2, leading: 12, bottom: 2, trailing: 12))
                    .listRowBackground(Color.clear)
                    .listRowSeparator(.hidden)
                    .swipeActions(edge: .trailing, allowsFullSwipe: false) {
                        Button(entry.dismissLabel, systemImage: "xmark") { dismiss(entry) }
                            .tint(colors.textSecondary)
                            .disabled(model.loading || model.dismissing.contains(entry.id))
                    }
                    .contextMenu {
                        Button(entry.dismissLabel, systemImage: "xmark") { dismiss(entry) }
                            .disabled(model.loading || model.dismissing.contains(entry.id))
                    }
                    .accessibilityAction(named: Text(entry.dismissLabel)) { dismiss(entry) }
                }
                if let error = model.error {
                    VStack(spacing: 12) {
                        Text(error).font(YappyFont.bodyMedium)
                        Button("Retry") { Task { await load() } }
                            .disabled(model.loading)
                    }
                    .frame(maxWidth: .infinity)
                    .listRowBackground(Color.clear)
                    .listRowSeparator(.hidden)
                } else if model.hasMore && model.loaded {
                    Button("Load more") { Task { await load() } }
                        .disabled(model.loading)
                        .frame(maxWidth: .infinity)
                        .listRowBackground(Color.clear)
                        .listRowSeparator(.hidden)
                        .onAppear { Task { await load() } }
                }
            }
            .listStyle(.plain)
            .scrollContentBackground(.hidden)
            .refreshable { await load(refresh: true) }
        }
        .frame(maxWidth: .infinity, maxHeight: .infinity)
        .neuBackdrop(colors)
        .navigationBarBackButtonHidden(true)
        .task { await load() }
    }

    @MainActor
    private func load(refresh: Bool = false) async {
        guard let userId = container.session.userId else { return }
        let generation = container.session.generation
        await model.load(container.repo, userId: userId, refresh: refresh,
                         isCurrent: { container.session.generation == generation })
        guard !Task.isCancelled else { return }
        await container.refreshBadge()
    }

    private func dismiss(_ entry: InboxEntry) {
        Task {
            await model.dismiss(entry, api: container.repo)
            await container.refreshBadge()
        }
    }

    private func row(_ entry: MentionEntry) -> some View {
        HStack(alignment: .top, spacing: 10) {
            /*
             * A mention still waiting for you.
             *
             * A bar down the leading edge and a tint behind the row, rather
             * than a bolder row: the list is already dense with names and
             * room titles, and making half of it heavier makes the whole
             * thing harder to scan. The bar is what the eye finds; the tint
             * says where the run ends.
             */
            if entry.unread {
                Capsule()
                    .fill(colors.accent)
                    .frame(width: 2, height: 36)
            }
            Avatar(
                url: entry.message?.sender?.avatarUrl,
                name: entry.message?.sender?.label,
                id: entry.message?.senderId ?? entry.conversation.id,
                size: 36
            )
            VStack(alignment: .leading, spacing: 2) {
                HStack(spacing: 6) {
                    // A channel names its space: "#general" alone is the title
                    // of half the channels anybody is in.
                    Text(entry.conversation.label)
                        .font(YappyFont.labelLarge)
                        .foregroundStyle(colors.textPrimary)
                        .lineLimit(1)
                    // A direct mention and a broadcast are not the same event
                    // to the person receiving one — somebody used your name, or
                    // you were in a room that got called.
                    if entry.isBroadcast {
                        Text("GROUP")
                            .font(YappyFont.labelSmall)
                            .foregroundStyle(colors.textTertiary)
                            .padding(.horizontal, 5)
                            .padding(.vertical, 1)
                            .background(colors.veil, in: RoundedRectangle(cornerRadius: 5))
                    }
                    Spacer(minLength: 0)
                    if let created = entry.message?.createdAt {
                        Text(YappyTime.relative(created))
                            .font(YappyFont.labelSmall)
                            .foregroundStyle(colors.textTertiary)
                    }
                }
                Text(entry.preview)
                    .font(YappyFont.bodyMedium)
                    .foregroundStyle(colors.textSecondary)
                    .lineLimit(2)
                    .multilineTextAlignment(.leading)
            }
        }
        .frame(maxWidth: .infinity, alignment: .leading)
        .padding(.horizontal, 12)
        .padding(.vertical, 10)
        /*
         * `accentSoft` rather than a fixed colour: it is defined per theme
         * — a pale violet on the light surface and a deep one on the dark —
         * so this is legible in both. A hardcoded rgba would have been right
         * in exactly one of them.
         */
        .background(
            entry.unread ? colors.accentSoft : Color.clear,
            in: NeuShape(radius: Neu.cornerMedium)
        )
        .contentShape(Rectangle())
        .softTap {
            guard let seq = entry.message?.seq else { return }
            onOpenMessage(entry.conversation.id, seq)
        }
    }
}

private extension MentionEntry {
    /// Stable across a reload; the message is what the row is about.
    var rowId: String { message?.id ?? conversation.id }

    var preview: String {
        let said = message?.content?.trimmingCharacters(in: .whitespacesAndNewlines)
        let body = (said?.isEmpty == false ? said! : "sent something")
        guard let who = message?.sender?.label else { return body }
        return "\(who)  \(body)"
    }
}

private extension MentionConversation {
    var label: String {
        let here = title ?? (type == "dm" ? "Direct message" : "Untitled")
        guard let parentTitle else { return here }
        return "\(parentTitle) / \(here)"
    }
}

File: android/app/src/main/java/gg/yappy/app/ui/chat/MentionsScreen.kt
package gg.yappy.app.ui.chat

import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.Spacer
import androidx.compose.foundation.layout.fillMaxSize
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.height
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.statusBarsPadding
import androidx.compose.foundation.layout.width
import androidx.compose.foundation.lazy.LazyColumn
import androidx.compose.foundation.lazy.items
import androidx.compose.material.icons.Icons
import androidx.compose.material.icons.automirrored.rounded.ArrowBack
import androidx.compose.material3.MaterialTheme
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.runtime.LaunchedEffect
import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.remember
import androidx.compose.runtime.setValue
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.draw.clip
import androidx.compose.ui.text.style.TextOverflow
import androidx.compose.ui.unit.dp
import androidx.compose.foundation.background
import androidx.compose.foundation.shape.RoundedCornerShape
import gg.yappy.app.data.MentionEntry
import gg.yappy.app.LocalContainer
import gg.yappy.app.ui.components.Avatar
import gg.yappy.app.ui.components.NeuIconButton
import gg.yappy.app.ui.components.softClickable
import gg.yappy.app.ui.theme.Neu
import gg.yappy.app.ui.theme.neuColors

/**
 * Everywhere you were called.
 *
 * One list across every group, so "where was I pinged while I was away" is a
 * question with an answer — before this it could only be reconstructed by
 * opening each room and looking for the badge, which is exactly the work a
 * notification list exists to save.
 */
@Composable
fun MentionsScreen(
    onBack: () -> Unit,
    /** Opens the room *at* the message, not merely at the bottom of it. */
    onOpenMessage: (conversationId: String, seq: Long) -> Unit,
) {
    val container = LocalContainer.current
    val colors = neuColors

    var entries by remember { mutableStateOf<List<MentionEntry>?>(null) }
    var failed by remember { mutableStateOf(false) }

    LaunchedEffect(Unit) {
        runCatching { container.repo.mentions().mentions }
            .onSuccess { entries = it }
            .onFailure { failed = true }
    }

    Column(
        Modifier
            .fillMaxSize()
            .background(colors.surface)
            // The screen draws edge to edge, so the header has to make room
            // for the clock rather than sitting under it.
            .statusBarsPadding(),
    ) {
        Row(
            Modifier.fillMaxWidth().padding(horizontal = 16.dp, vertical = 12.dp),
            verticalAlignment = Alignment.CenterVertically,
        ) {
            NeuIconButton(Icons.AutoMirrored.Rounded.ArrowBack, "Back", onBack)
            Spacer(Modifier.width(12.dp))
            Text(
                "Mentions",
                // headlineSmall, not titleLarge: the app's type scale does not
                // define titleLarge at all, so it fell through to Material's
                // default font and this one title was set in Roboto.
                style = MaterialTheme.typography.headlineSmall,
                color = colors.textPrimary,
            )
        }

        when {
            failed -> Empty("Couldn't load your mentions.")

            entries == null -> Empty("Loading…")

            entries!!.isEmpty() -> Empty(
                "Nobody has called you yet. When somebody uses your name, a role you " +
                    "hold, or @everyone, it lands here.",
            )

            else -> LazyColumn(
                Modifier.fillMaxSize(),
                contentPadding = androidx.compose.foundation.layout.PaddingValues(
                    horizontal = 12.dp,
                    vertical = 6.dp,
                ),
                verticalArrangement = Arrangement.spacedBy(4.dp),
            ) {
                items(entries!!, key = { it.message?.id ?: it.conversation.id }) { entry ->
                    MentionRow(entry) {
                        val seq = entry.message?.seq ?: return@MentionRow
                        onOpenMessage(entry.conversation.id, seq)
                    }
                }
            }
        }
    }
}

@Composable
private fun Empty(text: String) {
    val colors = neuColors
    Box(Modifier.fillMaxWidth().padding(32.dp), contentAlignment = Alignment.Center) {
        Text(
            text,
            style = MaterialTheme.typography.bodyMedium,
            color = colors.textTertiary,
        )
    }
}

@Composable
private fun MentionRow(entry: MentionEntry, onClick: () -> Unit) {
    val colors = neuColors
    val sender = entry.message?.sender

    Row(
        Modifier
            .fillMaxWidth()
            .clip(RoundedCornerShape(Neu.CornerMedium))
            /*
             * A mention still waiting for you.
             *
             * A tint plus a bar down the leading edge rather than a bolder
             * row: the list is already dense with names and room titles, and
             * making half of it heavier makes the whole thing harder to scan.
             * The bar is what the eye finds; the tint says where the run ends.
             *
             * `accentSoft` rather than a hardcoded colour, because it is
             * defined per theme — a pale violet on the light surface and a
             * deep one on the dark. A fixed rgba would have been legible in
             * exactly one of them.
             */
            .then(if (entry.unread) Modifier.background(colors.accentSoft) else Modifier)
            .softClickable(onClick = onClick)
            .padding(horizontal = 12.dp, vertical = 10.dp),
    ) {
        if (entry.unread) {
            Box(
                Modifier
                    .width(2.dp)
                    .height(36.dp)
                    .clip(RoundedCornerShape(Neu.CornerPill))
                    .background(colors.accent),
            )
            Spacer(Modifier.width(8.dp))
        }
        Avatar(
            url = sender?.avatarUrl,
            name = sender?.label,
            id = entry.message?.senderId ?: entry.conversation.id,
            size = 36.dp,
        )
        Spacer(Modifier.width(10.dp))
        Column(Modifier.fillMaxWidth()) {
            Row(verticalAlignment = Alignment.CenterVertically) {
                Text(
                    // A channel names its space: "#general" alone is the title
                    // of half the channels anybody is in.
                    entry.conversation.parentTitle?.let { "$it / ${entry.conversation.title ?: "channel"}" }
                        ?: (entry.conversation.title ?: "Direct message"),
                    style = MaterialTheme.typography.labelLarge,
                    color = colors.textPrimary,
                    maxLines = 1,
                    overflow = TextOverflow.Ellipsis,
                    modifier = Modifier.weight(1f, fill = false),
                )
                // A direct mention and a broadcast are not the same event to
                // the person receiving one — somebody used your name, or you
                // were in a room that got called.
                if (entry.isBroadcast) {
                    Spacer(Modifier.width(6.dp))
                    Text(
                        "GROUP",
                        style = MaterialTheme.typography.labelSmall,
                        color = colors.textTertiary,
                        modifier = Modifier
                            .clip(RoundedCornerShape(5.dp))
                            .background(colors.veil)
                            .padding(horizontal = 5.dp, vertical = 1.dp),
                    )
                }
            }
            Spacer(Modifier.height(2.dp))
            Text(
                buildString {
                    sender?.label?.let { append(it).append("  ") }
                    append(entry.message?.content?.trim()?.takeIf { it.isNotEmpty() } ?: "sent something")
                },
                style = MaterialTheme.typography.bodyMedium,
                color = colors.textSecondary,
                maxLines = 2,
                overflow = TextOverflow.Ellipsis,
            )
        }
    }
}
